azurelinuxagent/common/cgroupstelemetry.py

```diff
@@ -13,47 +13,184 @@
 
     @staticmethod
-    def is_tracked(name):
+    def is_tracked(name, controller):
         """
         Returns true if the given item is in the list of tracked items
         """
-        pass  # TODO
+        return name in CGroupsTelemetry._tracked and controller in CGroupsTelemetry._tracked[name]
 
     @staticmethod
     def stop_tracking(name):
         """
         Stop tracking the cgroups for the given name
         """
-        if CGroupsTelemetry.is_tracked(name):
-            del (CGroupsTelemetry._tracked[name])  # TODO - thread-safeness; review implementation
+        # Pop is atomic and thread-safe.
+        return CGroupsTelemetry._tracked.pop(name, None)
 
     @staticmethod
     def collect_all_tracked():
-        pass  # TODO
+        collected_metrics = {}
+        for name, cgroup_metrics in CGroupsTelemetry._cgroup_metrics.items():
+            collected_metrics[name] = CGroupsTelemetry._process_cgroup_metric(cgroup_metrics)
+            cgroup_metrics.clear()
+
+        return collected_metrics
+
+    @staticmethod
+    def poll_all_tracked():
+        CGroupsTelemetry._rlock.acquire()
+        for cgroup in CGroupsTelemetry._tracked:
+            if cgroup.name not in CGroupsTelemetry._cgroup_metrics:
+                CGroupsTelemetry._cgroup_metrics[cgroup.name] = CgroupMetrics()
+
+            CGroupsTelemetry._cgroup_metrics[cgroup.name].add_new_data(cgroup.controller,
+                                                                       cgroup.collect())
+            if not cgroup.is_active():
+                CGroupsTelemetry._tracked.remove(cgroup)
+
+        CGroupsTelemetry._rlock.release()
+
+    @staticmethod
+    def prune_all_tracked():
+        CGroupsTelemetry._rlock.acquire()
+        for cgroup in CGroupsTelemetry._tracked:
+            if not cgroup.is_active():
+                CGroupsTelemetry._tracked.remove(cgroup)
+
+        CGroupsTelemetry._rlock.release()
+
+    @staticmethod
+    def cleanup():
+        CGroupsTelemetry._tracked *= 0  # emptying the list
 
     def __init__(self):
-        pass  # TODO
+        pass
+        # TODO
+
+
+class CgroupMetrics(object):
+    def __init__(self):
+        self.current_memory_usage = Metric()
+        self.max_memory_levels = Metric()
+        self.current_cpu_usage = Metric()
+
+    def _add_memory_usage(self, metric):
+        self.current_memory_usage.append(metric[0].value)
+        self.max_memory_levels.append(metric[1].value)
+
+    def _add_cpu_usage(self, metric):
+        self.current_cpu_usage.append(metric[0].value)
 
+    def add_new_data(self, controller, metric):
+        if controller is "cpu":
+            self._add_cpu_usage(metric)
+        elif controller is "memory":
+            self._add_memory_usage(metric)
+
+    def get_metrics(self):
+        return self.current_memory_usage, self.max_memory_levels, self.current_cpu_usage
+
+    def clear(self):
+        self.current_memory_usage.clear()
```

Comment:
`clear()` isn't supported for lists in Python <3.3

Reply:
This clear calls the clear within Metric() (which internally does *= 0)